cmake/PythonExtensionHelpers.cmake

```diff
@@ -0,0 +1,115 @@
+# Python extension; rather, we simply produce a .so file with the correct name exported, and ensure
+# it's in the PYTHONPATH when testing.
+#
+# In our build files here, we build both kinds of extension for every Generator in the generators/
+# directory (even though not all are used). As a simplistic way to distinguish between the two
+# sorts of extensions, we use the unadorned Generator name for AOT extensions, and the Generator name
+# suffixed with "_stub" for Stub extensions. (TODO: this is unsatisfyingly hackish; better suggestions
+# would be welcome.)
+
+function(target_export_single_symbol TARGET SYMBOL)
+    configure_file("${_STUB_DIR}/ext.ldscript.apple.in" "${CMAKE_CURRENT_BINARY_DIR}/${TARGET}.ldscript.apple")
+    configure_file("${_STUB_DIR}/ext.ldscript.linux.in" "${CMAKE_CURRENT_BINARY_DIR}/${TARGET}.ldscript")
+    target_export_script(
+        ${TARGET}
+        APPLE_LD "${CMAKE_CURRENT_BINARY_DIR}/${TARGET}.ldscript.apple"
+        GNU_LD "${CMAKE_CURRENT_BINARY_DIR}/${TARGET}.ldscript"
+    )
+endfunction()
+
+function(add_python_aot_extension TARGET)
+    set(options)
+    set(oneValueArgs GENERATOR FUNCTION_NAME)
+    set(multiValueArgs SOURCES LINK_LIBRARIES FEATURES PARAMS)
+    cmake_parse_arguments(ARG "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    if (NOT ARG_GENERATOR)
+        set(ARG_GENERATOR "${TARGET}")
+    endif ()
+
+    if (NOT ARG_FUNCTION_NAME)
+        set(ARG_FUNCTION_NAME "${ARG_GENERATOR}")
+    endif ()
+
+    # Create the Halide generator executable.
+    add_executable(${TARGET}.generator ${ARG_SOURCES})
+    target_link_libraries(${TARGET}.generator PRIVATE Halide::Generator ${ARG_LINK_LIBRARIES})
+
+    # TODO: this should work (and would be preferred to the code above)
+    # but CMake fails with "targets not yet defined"; investigate.
+    # add_halide_generator(${TARGET}.generator
+    #                      SOURCES ${ARG_SOURCES})
+
+    # Run the Generator to produce a static library of AOT code,
+    # plus the 'python_extension' code necessary to produce a useful
+    # AOT Extention for Python:
+    add_halide_library(aot_${TARGET}
+                       FROM ${TARGET}.generator
+                       GENERATOR ${ARG_GENERATOR}
+                       FUNCTION_NAME ${ARG_FUNCTION_NAME}
+                       PYTHON_EXTENSION ${TARGET}.py.cpp
+                       FEATURES ${ARG_FEATURES}
+                       PARAMS ${ARG_PARAMS}
+                       TARGETS cmake)
+
+    # Take the native-code output of the Generator, add the Python-Extension
+    # code (to make it callable from Python), and build it into the AOT Extension we need.
+    Python3_add_library(${TARGET} MODULE ${${TARGET}.py.cpp})
+    target_link_libraries(${TARGET} PRIVATE aot_${TARGET})
+    set_target_properties(${TARGET} PROPERTIES OUTPUT_NAME ${ARG_GENERATOR}) # Python3_add_library adds target info to name.
```

Comment:
If we end up releasing this module to users, we will probably want a way to keep that info in the output name. IIUC, it's there to prevent the Python interpreter from loading an ABI-incompatible extension.

Reply:
Oops! I have to admit that I just cargo-culted this from previous code. Let's try removing it and seeing if things still work?

Reply:
I probably added it in there earlier to work around something I didn't know I was doing wrong :upside_down_face:

Reply:
...it may be that actually using `distutils`/`setuptools` instead is the right answer, but it felt kinda heavyweight for the needs here. That said, maybe we'd be better off using it as a way to Set A Good Example? I dunno.

Reply:
FYI, simply removing line 80 makes things fail to load. What (exactly) is Python3_add_library doing to the name? It seems to be poorly documented...

Reply:
https://github.com/Kitware/CMake/blob/eea53ea0a919b3c3f444433106ef578fcc429e5f/Modules/FindPython/Support.cmake#L3311-L3356

Reply:
It runs the equivalent of `python3-config --extension-suffix` to determine the SOABI version (on my system this is `cpython-38-x86_64-linux-gnu`) and then prepends `.${SOABI}` to the usual suffix (e.g. `.so`, `.pyd`).

Reply:
Also relevant: https://peps.python.org/pep-3149/

Reply:
Turns out this comment is misleading... the code itself is fine. It's getting rid of the `py_aot_` prefix from the target name. But CMake isn't actually even adding the soabi info in the first place.